File: bindings/python/tests/test_geotag.py
import pytest
import shutil
import os

from pylibuuas.geotag import getImageGeotag, geotagImage
from pylibuuas.telem import TelemetryArgs


@pytest.fixture
def test_telemetryArgs():
    targs = TelemetryArgs(
        latitude=1.0,
        longitude=2.0,
        altitude_agl_m=3.0,
        altitude_msl_m=4.0,
        heading_deg=5.0,
        velocityx_m_s=6.0,
        velocityy_m_s=7.0,
        velocityz_m_s=8.0,
        roll_rad=9.0,
        pitch_rad=10.0,
        yaw_rad=11.0,
        rollspeed_rad_s=12.0,
        pitchspeed_rad_s=13.0,
        yawspeed_rad_s=14.0,
        timestamp_pixhawk_ms=15,
        timestamp_msg_ms=16,
    )
    return targs


@pytest.fixture
def tagged_image_filename():
    SRC_IMAGE_FILENAME = "tests/resources/images/tagged_image1.jpg"
    TARGET_IMAGE_FILENAME = "/tmp/tests/resources/images/tagged_image1.jpg"
    os.makedirs("/tmp/tests/resources/images/", exist_ok=True)
    shutil.copy2(SRC_IMAGE_FILENAME, TARGET_IMAGE_FILENAME)
    return TARGET_IMAGE_FILENAME


@pytest.fixture
def untagged_image_filename():
    SRC_IMAGE_FILENAME = "tests/resources/images/image1.jpg"
    TARGET_IMAGE_FILENAME = "/tmp/tests/resources/images/untagged_image1.jpg"
    os.makedirs("/tmp/tests/resources/images/", exist_ok=True)
    shutil.copy2(SRC_IMAGE_FILENAME, TARGET_IMAGE_FILENAME)
    return TARGET_IMAGE_FILENAME


def test_getImageGeotag(tagged_image_filename):
    telem = getImageGeotag(tagged_image_filename)
    assert telem.latitude == 1.0
    assert telem.longitude == 2.0
    assert telem.altitude_agl_m == 3.0
    assert telem.altitude_msl_m == 4.0
    assert telem.heading_deg == 5.0
    assert telem.velocityx_m_s == 6.0
    assert telem.velocityy_m_s == 7.0
    assert telem.velocityz_m_s == 8.0
    assert telem.roll_rad == 9.0
    assert telem.pitch_rad == 10.0
    assert telem.yaw_rad == 11.0
    assert telem.rollspeed_rad_s == 12.0
    assert telem.pitchspeed_rad_s == 13.0
    assert telem.yawspeed_rad_s == 14.0
    assert telem.timestamp_pixhawk_ms == 15
    assert telem.timestamp_msg_ms == 16


def test_geotagImage(untagged_image_filename, test_telemetryArgs):
    result = geotagImage(untagged_image_filename, test_telemetryArgs)
    assert result == True


def test_imageGeotagThenGetImageGeotag(untagged_image_filename, test_telemetryArgs):
    result = geotagImage(untagged_image_filename, test_telemetryArgs)
    assert result == True

    telem = getImageGeotag(untagged_image_filename)
    assert telem.latitude == test_telemetryArgs.latitude
    assert telem.longitude == test_telemetryArgs.longitude
    assert telem.altitude_agl_m == test_telemetryArgs.altitude_agl_m
    assert telem.altitude_msl_m == test_telemetryArgs.altitude_msl_m
    assert telem.heading_deg == test_telemetryArgs.heading_deg
    assert telem.velocityx_m_s == test_telemetryArgs.velocityx_m_s
    assert telem.velocityy_m_s == test_telemetryArgs.velocityy_m_s
    assert telem.velocityz_m_s == test_telemetryArgs.velocityz_m_s
    assert telem.roll_rad == test_telemetryArgs.roll_rad
    assert telem.pitch_rad == test_telemetryArgs.pitch_rad
    assert telem.yaw_rad == test_telemetryArgs.yaw_rad
    assert telem.rollspeed_rad_s == test_telemetryArgs.rollspeed_rad_s
    assert telem.pitchspeed_rad_s == test_telemetryArgs.pitchspeed_rad_s
    assert telem.yawspeed_rad_s == test_telemetryArgs.yawspeed_rad_s
    assert telem.timestamp_pixhawk_ms == test_telemetryArgs.timestamp_pixhawk_ms
    assert telem.timestamp_msg_ms == test_telemetryArgs.timestamp_msg_ms

File: lib/test/test_geotag.cpp
#define private public

#include "test_utils.hpp"
#include <catch2/catch.hpp>
#include <libuuas/geotag.hpp>

using namespace libuuas::geotag;

constexpr char testImageFilenameJPEG[] = "test/resources/images/image1.jpg";
constexpr char testImageFilenamePNG[] = "test/resources/images/image1.png";
constexpr char testImageFilenameWEBP[] = "test/resources/images/image1.webp";

TEST_CASE("Geotag should successfully tag telemetry data", "[geotag]") {
    uuaspb::GeoTagRequest request;
    request.set_filename(testImageFilenameJPEG);
    request.mutable_telemetry()->set_latitude_dege7(1);
    request.mutable_telemetry()->set_longitude_dege7(2);
    request.mutable_telemetry()->set_altitude_agl_m(3);
    request.mutable_telemetry()->set_altitude_msl_m(4);
    request.mutable_telemetry()->set_heading_deg(5);
    request.mutable_telemetry()->set_velocityx_m_s(6);
    request.mutable_telemetry()->set_velocityy_m_s(7);
    request.mutable_telemetry()->set_velocityz_m_s(8);
    request.mutable_telemetry()->set_roll_rad(9);
    request.mutable_telemetry()->set_pitch_rad(10);
    request.mutable_telemetry()->set_yaw_rad(11);
    request.mutable_telemetry()->set_rollspeed_rad_s(12);
    request.mutable_telemetry()->set_pitchspeed_rad_s(13);
    request.mutable_telemetry()->set_yawspeed_rad_s(14);
    request.mutable_telemetry()->set_timestamp_pixhawk_ms(15);
    request.mutable_telemetry()->set_timestamp_msg_ms(16);
    uuaspb::GeoTagResponse response = geotagImage(request);
    REQUIRE(response.result() == uuaspb::ResultStatus::OK);

    uuaspb::UnGeoTagRequest getRequest;
    getRequest.set_filename(testImageFilenameJPEG);
    uuaspb::UnGeoTagResponse getResponse = getImageGeotag(getRequest);
    REQUIRE(getResponse.result() == uuaspb::ResultStatus::OK);
    REQUIRE(getResponse.telemetry().latitude_dege7() == request.telemetry().latitude_dege7());
    REQUIRE(getResponse.telemetry().longitude_dege7() == request.telemetry().longitude_dege7());
    REQUIRE(getResponse.telemetry().altitude_agl_m() == request.telemetry().altitude_agl_m());
    REQUIRE(getResponse.telemetry().altitude_msl_m() == request.telemetry().altitude_msl_m());
    REQUIRE(getResponse.telemetry().heading_deg() == request.telemetry().heading_deg());
    REQUIRE(getResponse.telemetry().velocityx_m_s() == request.telemetry().velocityx_m_s());
    REQUIRE(getResponse.telemetry().velocityy_m_s() == request.telemetry().velocityy_m_s());
    REQUIRE(getResponse.telemetry().velocityz_m_s() == request.telemetry().velocityz_m_s());
    REQUIRE(getResponse.telemetry().roll_rad() == request.telemetry().roll_rad());
    REQUIRE(getResponse.telemetry().pitch_rad() == request.telemetry().pitch_rad());
    REQUIRE(getResponse.telemetry().yaw_rad() == request.telemetry().yaw_rad());
    REQUIRE(getResponse.telemetry().rollspeed_rad_s() == request.telemetry().rollspeed_rad_s());
    REQUIRE(getResponse.telemetry().pitchspeed_rad_s() == request.telemetry().pitchspeed_rad_s());
    REQUIRE(getResponse.telemetry().yawspeed_rad_s() == request.telemetry().yawspeed_rad_s());
    REQUIRE(getResponse.telemetry().timestamp_pixhawk_ms() == request.telemetry().timestamp_pixhawk_ms());
    REQUIRE(getResponse.telemetry().timestamp_msg_ms() == request.telemetry().timestamp_msg_ms());
}

TEST_CASE("Geotag should successfully geotag a JPEG image", "[geotag]") {
    uuaspb::GeoTagRequest request;
    request.set_filename(testImageFilenameJPEG);

    uuaspb::GeoTagResponse response = geotagImage(request);
    REQUIRE(response.result() == uuaspb::ResultStatus::OK);
}

TEST_CASE("Geotag should successfully geotag a PNG image", "[geotag]") {
    uuaspb::GeoTagRequest request;
    request.set_filename(testImageFilenamePNG);

    uuaspb::GeoTagResponse response = geotagImage(request);
    REQUIRE(response.result() == uuaspb::ResultStatus::OK);
}

TEST_CASE("Geotag should successfully geotag a WEBP image", "[geotag]") {
    uuaspb::GeoTagRequest request;
    request.set_filename(testImageFilenameWEBP);

    uuaspb::GeoTagResponse response = geotagImage(request);
    REQUIRE(response.result() == uuaspb::ResultStatus::OK);
}
